driver/src/main/com/mongodb/Mongo.java

```diff
@@ -767,23 +767,67 @@ OperationExecutor createOperationExecutor() {
     }
 
     <T> T execute(final ReadOperation<T> operation, final ReadPreference readPreference) {
-        ReadBinding binding = getReadBinding(readPreference);
+        final ReadBinding binding = getReadBinding(readPreference);
         try {
-            return operation.execute(binding);
+            return doRetryWithExponentialBackoff(new BasicCallback<T>() {
+                @Override
+                public T execute() {
+                    return operation.execute(binding);
+                }
+            });
         } finally {
             binding.release();
         }
     }
 
     <T> T execute(final WriteOperation<T> operation) {
-        WriteBinding binding = getWriteBinding();
+        final WriteBinding binding = getWriteBinding();
         try {
-            return operation.execute(binding);
+            return doRetryWithExponentialBackoff(new BasicCallback<T>() {
+                @Override
+                public T execute() {
+                    return operation.execute(binding);
+                }
+            });
         } finally {
             binding.release();
         }
     }
 
+    <T> T doRetryWithExponentialBackoff(final BasicCallback<T> mainAttempt) {
+        int failure = 0;
+        int maxFailure = 6;
```

Comment:
At least move this to a static final in the class.

Reply:
I actually prefer putting our code at the same spot :(, I know generically they should be associated with class instead :)